src/rust/catloop/futures/connect.rs

```diff
@@ -0,0 +1,227 @@
+
+/// Client-side states in the connection establishment protocol.
+enum ClientState {
+    InitiateConnectRequest,
+    ConnectRequestSent(QToken),
+    ConnectAckReceived(QToken),
+    Connected(SocketAddrV4, Rc<DuplexPipe>),
+}
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// Descriptor for connect operation.
+pub struct ConnectFuture {
+    /// Underlying Catmem LibOS.
+    catmem: Rc<RefCell<CatmemLibOS>>,
+    /// Remote IPv4 address.
+    ipv4: Ipv4Addr,
+    /// Control duplex pipe used for establishing a the connection.
+    control_duplex_pipe: Rc<DuplexPipe>,
+    // State in the connection establishment protocol.
+    state: ClientState,
+}
+
+//======================================================================================================================
+// Associate Functions
+//======================================================================================================================
+
+/// Associate Functions for Connect Operation Descriptors
+impl ConnectFuture {
+    /// Creates a descriptor for a push operation.
+    pub fn new(catmem: Rc<RefCell<CatmemLibOS>>, remote: SocketAddrV4) -> Result<Self, Fail> {
+        let ipv4: &Ipv4Addr = remote.ip();
+        let port: u16 = remote.port().into();
+        let control_duplex_pipe: Rc<DuplexPipe> = Rc::new(DuplexPipe::open_duplex_pipe(catmem.clone(), ipv4, port)?);
+
+        Ok(ConnectFuture {
+            catmem,
+            ipv4: ipv4.clone(),
+            control_duplex_pipe,
+            state: ClientState::InitiateConnectRequest,
+        })
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+/// Future Trait Implementation for Connect Operation Descriptors
+impl Future for ConnectFuture {
+    type Output = Result<(SocketAddrV4, Rc<DuplexPipe>), Fail>;
+
+    /// Polls the target [ConnectFuture].
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+        let self_: &mut ConnectFuture = self.get_mut();
+
+        // Poll Catmem LibOS to make progress on ongoing operations.
+        self_.catmem.borrow_mut().poll();
+
+        // Act according to the state in the connection establishment protocol.
+        //
+        //  CLIENT                                                       SERVER
+        //    InitiateConnectRequest   --- ( msg: connect-request ) --->   WaitingConnectRequest
+        //    InitiateConnectRequest                                       WaitingConnectRequest
```

Comment:
Is this line correct?  Shouldn't the Client be in "ConnectRequestSent" immediately after sending the connect request?
And the Server go to "AcceptedConnectRequest"?

Reply:
Yes, and that is what I tried to draw using the lines. I intentionally replicated the states, to give some idea of time moving forward using the vertical axis. Note that state transitions are always indicated with vertical and horizontal arrows.

Reply:
Since line 108 comes after line 107 in time, any transitions resulting from the actions of line 107 should be reflected in the states of the client and server by line 108.

Reply:
I have changed the drawing based on what I understood that you want me to do. If the revised version does not meet your expectations, could you pate here what you have in mind?

Reply:
I'm still not seeing the changes to the drawings, so I don't know what's going on with GitHub.